crates/matrix-sdk/src/event_cache/linked_chunk.rs

```diff
@@ -0,0 +1,1036 @@
+    {
+        let items = items.into_iter();
+        let number_of_items = items.len();
+
+        let last_chunk = unsafe { self.last.as_mut().unwrap_or(&mut self.first).as_mut() };
+
+        // Push the items.
+        let last_chunk = last_chunk.push_items(items);
+
+        assert!(last_chunk.is_last_chunk(), "`last_chunk` must be… the last chunk");
+
+        // We need to update `self.last` if and only if `last_chunk` _is not_ the first
+        // chunk, and _is_ the last chunk (ensured by the `assert!` above).
+        if last_chunk.is_first_chunk().not() {
+            // Maybe `last_chunk` is the same as the previous `self.last` chunk, but it's
+            // OK.
+            self.last = Some(NonNull::from(last_chunk));
+        }
+
+        self.length += number_of_items;
+    }
+
+    /// Push a gap at the end of the [`LinkedChunk`], i.e. after the last
+    /// chunk.
+    pub fn push_gap_back(&mut self) {
+        let last_chunk = unsafe { self.last.as_mut().unwrap_or(&mut self.first).as_mut() };
+        last_chunk.insert_next(Chunk::new_gap_leaked());
+
+        self.last = last_chunk.next;
+    }
+
+    /// Insert items at a specified position in the [`LinkedChunk`].
+    ///
+    /// Because the `position` can be invalid, this method returns a
+    /// `Result`.
+    pub fn insert_items_at<I>(
+        &mut self,
+        items: I,
+        position: ItemPosition,
+    ) -> Result<(), LinkedChunkError>
+    where
+        I: IntoIterator<Item = T>,
+        I::IntoIter: ExactSizeIterator,
+    {
+        let items = items.into_iter();
+
+        let chunk_index = position.chunk_index();
+        let item_index = position.item_index();
+        let number_of_items = items.len();
+
+        let chunk = self
+            .nth_chunk_mut(chunk_index)
+            .ok_or(LinkedChunkError::InvalidChunkIndex { index: chunk_index })?;
+
+        let chunk = match &mut chunk.content {
+            ChunkContent::Gap => return Err(LinkedChunkError::ChunkIsAGap { index: chunk_index }),
+            ChunkContent::Items(current_items) => {
+                let current_items_length = current_items.len();
+
+                if item_index >= current_items_length {
+                    return Err(LinkedChunkError::InvalidItemIndex { index: item_index });
+                }
+
+                // The `ItemPosition` is computed from the latest items. Here, we manipulate the
+                // items in their original order: the last item comes last. Let's adjust
+                // `item_index`.
+                let item_index = current_items_length - 1 - item_index;
+
+                // Split the items.
+                let detached_items = current_items.split_off(item_index);
+
+                chunk
+                    // Push the new items.
+                    .push_items(items)
+                    // Finally, push the items that have been detached.
+                    .push_items(detached_items.into_iter())
```

Comment:
Why?

Reply:
Pushing after the last item would be quite frequent, I thought it would avoid heap allocations but maybe in this case the `split_off` doesn't allocate since the returned `Vec` is empty, so that might not avoid any allocations?

Reply:
There is an allocation for the returned vector whatever, https://github.com/rust-lang/rust/blob/e69f14b14cfa6e2ae863adb07e9add736d1f45fc/library/alloc/src/vec/mod.rs#L2276.